File: app/src/main/java/com/univreview/fragment/upload/UploadReviewDetailFragment.java
package com.univreview.fragment.upload;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.univreview.App;
import com.univreview.Navigator;
import com.univreview.R;
import com.univreview.activity.BaseActivity;
import com.univreview.fragment.BaseWriteFragment;
import com.univreview.fragment.review.ReviewDetailFragment;
import com.univreview.log.Logger;
import com.univreview.model.Review;
import com.univreview.model.ReviewDetail;
import com.univreview.network.Retro;
import com.univreview.util.Util;
import com.univreview.view.ReviewItemView;

import butterknife.BindView;
import butterknife.ButterKnife;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by DavidHa on 2017. 1. 13..
 */
public class UploadReviewDetailFragment extends BaseWriteFragment {
    private static final int POSITION_NONE = -1;
    @BindView(R.id.review_item) ReviewItemView reviewItemView;
    @BindView(R.id.input_review) EditText inputReview;
    private boolean isUpdate = false;
    private Review review;
    private ReviewDetail reviewDetail;

    public static UploadReviewDetailFragment newInstance(Review review){
        UploadReviewDetailFragment fragment = new UploadReviewDetailFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable("review", review);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        review = (Review) getArguments().getSerializable("review");
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_upload_review_detail, container, false);
        ButterKnife.bind(this, view);
        setReviewData(review);
        toolbar.setToolbarBackgroundColor(R.color.colorPrimary);
        toolbar.setBackBtnVisibility(true);
        toolbar.setOnConfirmListener(v -> registerReview(review.id));
        rootLayout.addView(view);
        if (review.reviewDetail != null) {
            isUpdate = true;
            toolbar.setTitleTxt("리뷰 수정");
            inputReview.setText(review.reviewDetail.reviewDetail);
            inputReview.setSelection(inputReview.getText().toString().length());
        }else{
            toolbar.setTitleTxt("상세리뷰 쓰기");
        }
        return rootLayout;
    }

    private void setReviewData(Review review) {
        reviewItemView.setData(review, POSITION_NONE);
        reviewItemView.setMode(ReviewItemView.Status.WRITE_REVIEW);
    }

    private void registerReview(long reviewId) {
        Util.hideKeyboard(context, inputReview);
        reviewDetail = new ReviewDetail();
        reviewDetail.reviewId = reviewId;
        reviewDetail.reviewDetail = inputReview.getText().toString();
        if (reviewDetail.getAlertMessage() == null) {
            Logger.v("review detail: " + reviewDetail);
            progressDialog.show();
            if(isUpdate){
                callPutReviewDetail(review.reviewDetail.id, reviewDetail);
            }else {
                callPostReviewDetail(reviewDetail);
            }
        } else {
            Util.simpleMessageDialog(context, reviewDetail.getAlertMessage());
        }
    }

    private void callPostReviewDetail(ReviewDetail reviewDetail) {
        Retro.instance.reviewService().postDetailReview(App.setAuthHeader(App.userToken), reviewDetail)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate(() -> progressDialog.dismiss())
                .subscribe(result -> {
                    ReviewDetailFragment.isRefresh = true;
                    Navigator.goReviewDetail(context, review);
                    ((BaseActivity) activity).setOnBackPressedListener(null);
                    activity.onBackPressed();
                }, this::errorResponse);
    }

    private void callPutReviewDetail(long reviewDetailId, ReviewDetail reviewDetail) {
        Logger.v("review detail id: " + reviewDetailId);
        Retro.instance.reviewService().putReviewDetail(App.setAuthHeader(App.userToken), reviewDetailId, reviewDetail)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate(() -> progressDialog.dismiss())
                .subscribe(result -> {
                    ReviewDetailFragment.isRefresh = true;
                    ((BaseActivity) activity).setOnBackPressedListener(null);
                    activity.onBackPressed();
                }, this::errorResponse);
    }


    private void errorResponse(Throwable e) {
        Logger.e(e);
    }


}

File: app/src/main/java/com/univreview/fragment/upload/UploadReviewDetailFragment.kt
package com.univreview.fragment.upload

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.univreview.App
import com.univreview.Navigator
import com.univreview.R
import com.univreview.fragment.BaseWriteFragment
import com.univreview.log.Logger
import com.univreview.model.enumeration.ReviewType
import com.univreview.model.Review
import com.univreview.model.ReviewDetail
import com.univreview.network.Retro
import com.univreview.util.Util
import kotlinx.android.synthetic.main.fragment_upload_review_detail.*
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by DavidHa on 2017. 8. 8..
 */
class UploadReviewDetailFragment : BaseWriteFragment() {
    companion object {

        @JvmStatic
        fun getInstance(review: Review, isFirst: Boolean): UploadReviewDetailFragment {
            val fragment = UploadReviewDetailFragment()
            val bundle = Bundle()
            bundle.putParcelable("review", review)
            bundle.putBoolean("isFirst", isFirst)
            fragment.arguments = bundle
            return fragment
        }
    }

    private var isFirst: Boolean = false
    private lateinit var review: Review


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        review = arguments.getParcelable("review")
        isFirst = arguments.getBoolean("isFirst")
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val view = inflater?.inflate(R.layout.fragment_upload_review_detail, container, false)
        toolbar.setToolbarBackgroundColor(R.color.colorPrimary)
        toolbar.setBackBtnVisibility(true)
        toolbar.setOnRightMenuListener { registerReview(review.id) }
        rootLayout.addView(view)
        return rootLayout
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    private fun init() {
        setReviewData(review)
        if (review.content != null) {
            toolbar.setTitleTxt("리뷰 수정")
            inputReview.setText(review.content)
            inputReview.setSelection(inputReview.text.toString().length)
        } else {
            toolbar.setTitleTxt("상세리뷰 쓰기")
        }
    }

    private fun setReviewData(review: Review) {
        reviewItem.setMode(ReviewType.WRITE_REVIEW)
        reviewItem.setData(review)
    }

    private fun registerReview(reviewId: Long) {
        Util.hideKeyboard(context, inputReview)
        val reviewDetail = ReviewDetail(inputReview.text.toString().trim())
        if (reviewDetail.getAlertMessage() == null) {
            Logger.v("review detail: " + reviewDetail)
            progressDialog.show()
            callPutReviewDetail(reviewId, reviewDetail)
        } else {
            Util.simpleMessageDialog(context, reviewDetail.getAlertMessage())
        }
    }

    // 리뷰 작성 경로
    // UploadReview -> UploadReviewDetail -> (ReviewDetail)
    // MyReview List -> ReviewDetail -> UploadReviewDetail


    // 리뷰 수정 경로
    // MyReview List -> ReviewDetail -> UploadReviewDetail

    private fun callPutReviewDetail(reviewId: Long, reviewDetail: ReviewDetail) {
        Logger.v("review id: " + reviewId)
        Retro.instance.reviewService().callPutReview(App.setHeader(), reviewId, reviewDetail)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate { progressDialog.dismiss() }
                .subscribe({
                    review.content = it.data.content
                    review.notifyUpdate()
                    if (isFirst) Navigator.goReviewDetail(context, review)
                    activity.finish()
                }, { this.errorResponse(it) })
    }


    private fun errorResponse(e: Throwable) {
        Logger.e(e)
    }
}